### DISCUSSION TITLE
Inner/Nested Class Within Function Return Type Not Defined

### DISCUSSION THREAD

**Discussion by jodhus** [python/typing#1834] [Q&A] [answered]
Apologies for the title word salad.

I have a static class factory function that I defined as returning the inner class; however, I'm getting mixed results with type checking:

```python
def class_factory() -> "InnerClass":
    class InnerClass:
        pass
    return InnerClass
```

Pylint seems to have no problem, but Mypy is complaining that "InnerClass" is not defined. I've tried `from __future__ import annotations`, but that didn't help either.

Is there a better way to define this or is this potentially a bug in Mypy? Thanks!

**Comment by erictraut** [comment, marked as answer]
Mypy is correct to generate an error here. Pylint is probably just missing a check.

The `InnerClass` symbol is defined in an inner scope and has no meaning where you are referencing it. The quotes mean that forward references are allowed, but this symbol is not a forward reference.

If you want the type declaration to be visible outside of the function, you should declare it outside of the function, like this:

```python
class MyClass:
    pass

def class_factory() -> type[MyClass]:
    return MyClass
```

**Reply by jodhus** [reply]
Funny enough, I normally don't do inner/nested classes. I should have stuck to my instincts.

Appreciate the insight. Thanks!
